spi/src/main/java/org/opensearch/jobscheduler/spi/utils/JobHistoryService.java

```diff
@@ -0,0 +1,263 @@
+                            listener.onFailure(exception);
+                        }
+                    }));
+            }
+        } catch (Exception e) {
+            logger.error(e);
+            listener.onFailure(e);
+        }
+    }
+
+    /**
+     * Records job execution history to the history index.
+     *
+     * @param startTime the time when job execution started
+     * @param endTime the time when job execution ended (can be null for ongoing jobs)
+     * @param status the execution status (e.g., "RUNNING", "SUCCESS", "FAILED")
+     * @param listener callback for handling the response
+     */
+    public void recordJobHistory(
+        final String jobIndexName,
+        final String jobId,
+        final Instant startTime,
+        final Instant endTime,
+        final Integer status,
+        ActionListener<Boolean> listener
+    ) {
+        if (jobIndexName == null || jobId == null || startTime == null) {
+            listener.onFailure(new IllegalArgumentException("JobIndexName, JobId, StartTime, and Status cannot be null"));
+            return;
+        }
+
+        createHistoryIndex(ActionListener.wrap(created -> {
+            if (created) {
+                findHistoryRecord(jobIndexName, jobId, startTime, ActionListener.wrap(existingRecord -> {
+                    if (existingRecord != null) {
+                        // Update existing record
+                        StatusHistoryModel updatedModel = new StatusHistoryModel(
+                            jobIndexName,
+                            jobId,
+                            startTime,
+                            endTime,
+                            status,
+                            existingRecord.getSeqNo(),
+                            existingRecord.getPrimaryTerm()
+                        );
+                        updateHistoryRecord(
+                            updatedModel,
+                            ActionListener.wrap(updated -> listener.onResponse(updated != null), listener::onFailure)
+                        );
+                    } else {
+                        // Create new record
+                        try {
+                            StatusHistoryModel historyModel = new StatusHistoryModel(jobIndexName, jobId, startTime, endTime, status);
+                            createHistoryRecord(historyModel, listener);
+                        } catch (Exception e) {
+                            logger.error("Failed to create history record", e);
+                            listener.onFailure(e);
+                        }
+                    }
+                }, listener::onFailure));
+            } else {
+                listener.onResponse(false);
+            }
+        }, listener::onFailure));
+    }
+
+    private void createHistoryRecord(final StatusHistoryModel historyModel, ActionListener<Boolean> listener) {
+        try (ThreadContext.StoredContext ignore = client.threadPool().getThreadContext().stashContext()) {
```

Comment:
we can skip stashing context here since it is only called inside findHistoryRecord which stashes context.